Sources/PackageRegistry/RegistryManager.swift

```diff
@@ -0,0 +1,253 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import PackageLoading
+import PackageModel
+
+import TSCBasic
+import TSCUtility
+
+import Foundation
+#if canImport(FoundationNetworking)
+import FoundationNetworking
+#endif
+
+public enum RegistryError: Error {
+    case invalidOperation
+    case invalidResponse
+    case invalidURL
+    case invalidChecksum(expected: String, actual: String)
+    case invalidPackage
+}
+
+public final class RegistryManager {
+    public static var configuration: URLSessionConfiguration = {
+        var configuration = URLSessionConfiguration.default
+        configuration.httpAdditionalHeaders = [
+            "User-Agent": "Swift Package Manager",
+        ]
+
+        return configuration
+    }()
+
+    private lazy var session = URLSession(configuration: Self.configuration)
+
+    let registryBaseURL: Foundation.URL
+
+    init(registryBaseURL: Foundation.URL) {
+        self.registryBaseURL = registryBaseURL
+    }
+
+    public class func discover(
+        for package: PackageReference,
+        timeoutInterval: TimeInterval = 1.0,
+        completion: @escaping (Result<RegistryManager, Error>) -> Void
+    ) {
+        guard case .remote(_, let urlString) = package,
+              urlString.hasPrefix("https://"),
+              !urlString.hasSuffix(".git")
+        else {
+            return completion(.failure(RegistryError.invalidPackage))
+        }
+
+        guard let url = URL(string: "https://\(package.identity)") else {
+            return completion(.failure(RegistryError.invalidURL))
+        }
+
+        var request = URLRequest(url: url, cachePolicy: .returnCacheDataElseLoad, timeoutInterval: timeoutInterval)
+        request.httpMethod = "HEAD"
+        request.addValue("application/vnd.swift.registry.v1", forHTTPHeaderField: "Accept")
```

Comment:
a few more things to address:
* retries  with incremental backoff
* circuit breaking when we detect a registry is in a bad state